### DISCUSSION TITLE
adding new constraints on the fly

### DISCUSSION THREAD

**Discussion by maolingao** [yalmip/YALMIP#909] [General discussion] [answered]
Hi Johan,

I am currently using yalmip to solve my MIP problem, so far yalmip works great.

Now our the MIP model changes a bit: we would like to add new constraints on-the-fly, which depends on the current solution of underlying solvers, say mosek/gurobi (via callbacks). I found some related articles on the callbacks (see below links), however very limited information on adding new constraints dynamically. 

I would like to ask, if you have experience related to this. And it would super nice and convenient, if there is a way to do this via yalmip. Unfortunately I didn’t find anything related. It will be very helpful, if you could give some pointers on this topic?

Thanks for your kind help in advance,
Maolin Gao

Callbacks:

https://docs.mosek.com/9.2/pythonapi/callback.html
https://www.gurobi.com/documentation/9.1/refman/cb_codes.html

**Comment by johanlofberg** [comment, marked as answer]
Last time I checked, neither mosek nor gurobi/xpress/cplex support callbacks in their top-level MATLAB interfaces which YALMIP uses.

A hack which can be used for smaller examples where you have an upper bound on the added constraints is to setup an optimizer object with constraints which are redundant unless data makes them active. It will not work for large-scale problems as the symbolic constraints become complex, but for examples like the following which is a variant of https://yalmip.github.io/example/scenariosampling/, it works well.

```
x = sdpvar(2,1);
Objective = -sum(x);
Ball = [-1 <= x <= 1];
plot(Ball);hold on

N = 200;
A = sdpvar(N,2);
Dynamic = [Ball, A*x <= 1];

Generated = zeros(N,2);
Solver = optimizer(Dynamic,Objective,sdpsettings('solver','mosek'),A,x)
for i = 1:N
    tic;Solver(Generated);toc
    v = randn(1,2);v = v/norm(v);
    Generated(i,:) = v;
end
plot(Solver(Generated,'nosolve'))   

```

**Comment by johanlofberg** [comment]
I had to do some experimentation to see if it is possible to add cuts without a predefined size. It is, using the sampler approach in the end of the linked article.

It's not what the function is built for, but a fun challenge. The idea is to create a sampler which returns samples from a saved list of data, adding data and then asking for more and more samples and constantly starting from the first added sample.

```
x = sdpvar(2,1);
Objective = -sum(x);
Ball = [-1 <= x <= 1];
plot(Ball);hold on

f = @(x)callback(x);
A = sdpvar(1,2);
% Populate with something so optimizer can test
callback([0 0],'reset');
Dynamic = [Ball, A*x <= 1,uncertain(A,f)];
Solver = optimizer(Dynamic,Objective,sdpsettings('solver','mosek'),A,x);
% Reset after optimizer tested
callback([0 0],'reset');
N = 5000;
for i = 1:N
    tic;Q = sample(Solver,i);toc
    plot(Q);drawnow
    x = Q();
    v = randn(1,2);v = v/norm(v);
    callback(v,'add');
end
plot(Solver(Generated,'nosolve'))

function v = callback(in,flag)
persistent saved index skip
if nargin == 2
    if strcmp(flag,'reset')
        index = 1;
        saved = in;
    else
        index = 1;
        saved = [saved;in];
    end
else
    v = saved(index,:);
    index = index + 1;
end
end
```

**Comment by maolingao** [comment]
great, the answer is very helpful.

One difference of the new constraints is that ours would depends on the current solution of the solver, say we check the solution every 10 solver iterations and based on the current solution, new constraints will be added. Is there a way to obtain the current solution? it seems hard, since mosek, gurobi and others don't support callback in matlab api.

thanks!

**Reply by johanlofberg** [reply]
If you mean a solution that the solver has returned in a call to the optimizer object, that just means you send and store something else (which depends on the solution) to the callback data holder instead of the random data I sent.

If you mean current internal iterate in the solver, well that is precisely what is impossible, since callbacks aren't supported.

**Reply by maolingao** [reply]
I meant the current internal iterate in the solver.

However when it comes to the solution that the solver has returned in a call to the optimizer object, based on which we add new constraints, then run the optimizer again. Would this actually rerun the whole MIP problem? (since it is costly to solve MIP problem, it will be very expensive to run it multiple times.)

**Reply by johanlofberg** [reply]
Yes, complete restart. You can ask the solver to try to warm-start from the last iterate computed in optimizer by setting the options usex0 to 1.

**Reply by maolingao** [reply]
true, this will help. However all solvers support warm start I think, f.e. mosek doesn't.

**Reply by johanlofberg** [reply]
in many problems a good intial solution, or even the optimal one, does not help at all, as the hard part is not to compute the optimal solution, but to prove optimality.

Mosek supports warm-start for integer programs (but for some reason that flag has gone missing in the setup of solver definitions, so I will have to fix that)

**Reply by johanlofberg** [reply]
I've added back integer warm-starts and pushed to the develop branch

**Reply by maolingao** [reply]
thanks!

After talking to you it becomes apparent to me: since the currently bottemneck doesn't lie in YALMIP, but the optimisation of underlying solvers, I would expect using the `optimizer` class won't save much time in the end, b/c the MIP solver has to restart every time after adding new constraints, and do the hard/time-intensive work of proving optimality. 

It seems what we would like to have is more an interruption of the underlying solver, so that the current iterate can be checked, and new constraints can be added on-demand, then resume the optimization from where we pause. I tend to think that the underlying mechanism is harder, since adding new constraints will also partially invalidate the knowledge on the bound accumulated up till now, which means the MIP solver has to update the whole exploration history.
